File: 백준/Gold/9935. 문자열 폭발/문자열 폭발.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Scanner;
import java.util.Stack;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    static String str;
    static String bomb;


    public static void main(String[] args) throws IOException {
        input();

        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < str.length(); i++) {
            char alpha = str.charAt(i);
            //stack에 넣기
            stack.add(alpha);

            //stack의 size가 bomb보다 크거나 같으면 비교
            int pointer = bomb.length() - 1;
            if (stack.size() >= bomb.length() && alpha==bomb.charAt(bomb.length()-1)) {
                Stack<Character> tmp = new Stack<>();

                for (int k = 0; k < bomb.length(); k++) {
                    char now = stack.pop();
                    tmp.add(now);
                    if (bomb.charAt(pointer) == now) {
                        pointer--;

                    } else {
                        while (!tmp.isEmpty()) {
                            char tmpAlph = tmp.pop();
                            stack.add(tmpAlph);
                        }
                    }
                }
            }


        }
        if (stack.isEmpty()) System.out.println("FRULA");
        else {
            StringBuilder sb = new StringBuilder();
            while (!stack.isEmpty()) {
                char now = stack.pop();
                sb.append(now);
            }

            System.out.println(sb.reverse().toString());
        }
    }

    public static void input() throws IOException {
        str = br.readLine();
        bomb = br.readLine();

    }
}

File: 백준/Gold/9935. 문자열 폭발/문자열 폭발.py
string = input()
bomb = input()
l = len(bomb)
stk = []
idx = 0
for s in string:
    #idx 값 세팅
    stk.append(s)
    if idx == l-1 and bomb[idx] == s :
        for k in range(l):
            stk.pop()
        idx = 0
    elif s in bomb:
        i = bomb.index(s)

        if len(stk) > i: #스택에 bomb 앞 문자열만큼의 문자가 들어있으면
            cnt = 0
            for k in range(i): #top부터 그 앞까지 문자열 일치하는지 확인
                if stk[-i-1+k] == bomb[k]:
                    cnt += 1
            if cnt == i:
                idx = i+1
                if idx == l:
                    for k in range(l):
                        stk.pop()
                    idx = 0
            else:
                idx = 0
        else:
            idx = 0
    else:
        idx = 0

if stk:
    print(*stk, sep='')
else:
    print("FRULA")


